pypesto/optimize/optimizer.py

```diff
@@ -41,6 +41,16 @@ def __init__(self, optimizer: str):
         )
 
 
+def add_inner_parameters(
+    objective: Objective, optimizer_result: OptimizerResult
+):
+    """Add inner parameters from objective to the optimizer result."""
+    if hasattr(objective, INNER_PARAMETERS) and any(
+        objective.inner_parameters
+    ):
+        optimizer_result[INNER_PARAMETERS] = objective.inner_parameters
```

Comment:
I don't really like the idea of the `Optimizer` being aware of hierarchical optimization. In my opinion, this should all be hidden inside the given `Problem`.
Instead of having an optimizer call `add_inner_parameters`, I'd prefer the optimizer calling sth like `Problem.postprocess_optimization_result`. That's also not great - open for better suggestions - because `Problem` shouldn't have to care about whether it's being optimizer, sampled, or whatever, but I'd still prefer it that way around.

Reply:
Ohhh I like the final comment idea you had. Renaming the minimize function of each individual optimizer `_minimize`, and the minimize call being called from the base (parent) object which would postprocess it as well. Definitely sounds nicer, and more consistent with respect to the current objective implementation (unprocessed call etc...). Would like to hear the opinions of others on what's best to do.

Reply:
One alternative is to implement this as a wrapper, like `history_decorator`. I agree with Daniel, it's not clear to me whether it's better for `Optimizer` to be aware of `inner_parameters`, or for `Problem` to be aware of `OptimizerResult`. Maybe "`Optimizer` knows `inner_parameters`" is less circular, since the `inner_parameters` are actually output from the objective, and `Optimizer` already knows `problem.objective`.

Somewhat related: since there are so many decorators applied to `minimize`, maybe a single decorator that collects all other decorators will make the code cleaner.

> Renaming the minimize function of each individual optimizer _minimize, and the minimize call being called from the base (parent) object which would postprocess it as well.

This sounds like an alternative to the currect decorators, but if the current decorators are replaced with this `_minimize`+`minimize` setup, then you would require a `preprocess` step too -- or, will you keep the current decorators, and just have this setup only for inner parameters? I think sticking to decorators is probably the cleanest option.

Does the current code also allow users to save inner parameters from profiling? If not, is this planned, and how would it be done?

Reply:
Decorators: As good.

> ...  or for `Problem` to be aware of `OptimizerResult`

I think at some point we might need something in that direction (e.g. plans for some PEtab result file). Maybe we can get around knowing about `OptimizerResult` specifically by making all `*Result` handle some kind of `ProblemInternals`.
This is probably out of scope here. I would be happy with a postprocessing solution in `Optimizer` (based on decorators or methods), so we have that handled in a single location. That should make pushing things from Optimizer to Problem much easier later on.

Reply:
Seems like the cleanest thing now to do would be to add it as a decorator and collapse all decorators into one. I'll try that now. If it doesn't fit well, I can change it to the `_minimize` + `minimize` option.

Reply:
Done, seems to be working nicely :)